README.md
# k8scontroller-metadata-exporter
![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/vfabi/k8scontroller-metadata-exporter)
![GitHub last commit](https://img.shields.io/github/last-commit/vfabi/k8scontroller-metadata-exporter)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

[![Generic badge](https://img.shields.io/badge/hub.docker.com-vfabi/k8scontroller_metadata_exporter-<>.svg)](https://hub.docker.com/repository/docker/vfabi/k8scontroller-metadata-exporter)
![Docker Cloud Automated build](https://img.shields.io/docker/cloud/automated/vfabi/k8scontroller-metadata-exporter)
![Docker Pulls](https://img.shields.io/docker/pulls/vfabi/k8scontroller-metadata-exporter)
![Docker Cloud Build Status](https://img.shields.io/docker/cloud/build/vfabi/k8scontroller-metadata-exporter)

Basic Kubernetes controller to get Kubernetes objects metadata (deployments, pods, etc) and export it via HTTP API endpoint as json.  
This web application acting as proxy that **reads only** data from Kubernetes API and provides it for the other external applications as a **reads only** access point to the Kubernetes object's metadata without direct access to Kubernetes API.

## Features
- Reads Kubernetes objects metadata for all namespaces from Kubernetes API and exports it as http/json. Supported methods:  
    - deployments list
    - pods list
    - pod logs
- Strict namespace to domain mapping - allows to output only metadata from namespace mapped to specific domain.
- Data filtering by namespace key.


# Technology stack
- Python 3.6+
- Flask
- kubernetes


# Requirements and dependencies
## Application
Python libs in requirements.txt

## External
Kubernetes instance and Nginx ingress controller installed.


# Configuration
## Environment variables
| Name | Required | Values | Description |
|:----------|:-------------:|:------|:------|
|STRICT_NAMESPACE_MAPPING|False||Strict namespace to domain mapping, example: `'frontend.develop.example.com:develop,frontend.staging.example.com:staging'`|


# Usage
1. Apply Kubernetes RBAC configuration:
```yaml
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: k8scontroller-metadata-exporter
  namespace: test

---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  namespace: test
  name: k8scontroller-metadata-exporter
rules:
- apiGroups: ["extensions", "apps"]
  resources: ["deployments", "replicasets"]
  verbs: ["get", "list"]
- apiGroups:
  - ""
  resources:
  - pods
  - pods/log
  verbs:
  - get
  - watch
  - list

---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: k8scontroller-metadata-exporter
  namespace: test
subjects:
- kind: ServiceAccount
  name: k8scontroller-metadata-exporter
  namespace: test
roleRef:
  kind: ClusterRole
  name: k8scontroller-metadata-exporter
  apiGroup: rbac.authorization.k8s.io
```

2. Apply k8scontroller-metadata-exporter K8S Deployment, Service, ConfigMap and Ingress:
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: k8scontroller-metadata-exporter
  namespace: test
data:
  # No special chars and spaces for value
  STRICT_NAMESPACE_MAPPING: frontend.develop.example.com:develop,frontend.staging.example.com:staging

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: k8scontroller-metadata-exporter
  namespace: test
spec:
  selector:
    matchLabels:
      app: k8scontroller-metadata-exporter
  replicas: 1
  template:
    metadata:
      labels:
        app: k8scontroller-metadata-exporter
    spec:
      serviceAccountName: k8scontroller-metadata-exporter
      containers:
      - name: k8scontroller-metadata-exporter
        image: vfabi/k8scontroller-metadata-exporter:latest
        ports:
        - containerPort: 8000
        envFrom:
        - configMapRef:
            name: k8scontroller-metadata-exporter

---
apiVersion: v1
kind: Service
metadata:
  name: k8scontroller-metadata-exporter
  namespace: test
spec:
  type: NodePort
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8000
  selector:
    app: k8scontroller-metadata-exporter

---
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: k8scontroller-metadata-exporter
  namespace: test
  annotations:
    kubernetes.io/ingress.class: "nginx"
    nginx.ingress.kubernetes.io/rewrite-target: /$2
spec:
  rules:
  - host: frontend.develop.example.com
    http:
      paths:
      - path: /kmeta(/|$)(.*)
        backend:
          serviceName: k8scontroller-metadata-exporter
          servicePort: 80
```

> **Note**:  
> Please change host value (frontend.develop.example.com) in the Ingress configuration to yours.  
IMPORTANT! This is a test Kubernetes configuration for k8scontroller-metadata-exporter. You should use HTTPS configuration for your Ingress.

You can access k8scontroller-metadata-exporter at http://frontend.develop.example.com/kmeta/deployments/  

### API endpoints
  - /deployments/ - get Kubernetes deployments list  
  Params, filtering (optional) example: http://frontend.develop.example.com/kmeta/deployments/?namespace=develop  
  If request domain is specified in strict namespace mapping (STRICT_NAMESPACE_MAPPING env variable) this filtering feature won't work.  

  - /pods/ - get Kubernetes pods list  
  Params, filtering (optional) example: http://frontend.develop.example.com/kmeta/pods/?namespace=develop  

  - /pod/logs/ - get Kubernetes specific pod logs  
  Params (required) example: http://frontend.develop.example.com/kmeta/pod/logs/?namespace=develop&pod=application-7fcf8df75d-pr545&tail_lines=100  

> **Note**:  
> Strict namespace mapping feature allow to map request domain only get Kubernetes objects metadata only from specified for it namespace.  
For example, you have 2 domains attached to Kubernetes frontend.develop.example.com and frontend.staging.example.com. You have configured Ingress and would like to provide access for application that serves requests at frontend.develop.example.com for Kubernetes objects metadata from develop namespace only - just put this data in STRICT_NAMESPACE_MAPPING env variable like this: `frontend.develop.example.com:develop`, or for 2 domains/namespaces: `frontend.develop.example.com:develop,frontend.staging.example.com:staging` accordingly.


# Docker
[![Generic badge](https://img.shields.io/badge/hub.docker.com-vfabi/k8scontroller_metadata_exporter-<>.svg)](https://hub.docker.com/repository/docker/vfabi/k8scontroller-metadata-exporter)


# Contributing
Please refer to each project's style and contribution guidelines for submitting patches and additions. In general, we follow the "fork-and-pull" Git workflow.

 1. **Fork** the repo on GitHub
 2. **Clone** the project to your own machine
 3. **Commit** changes to your own branch
 4. **Push** your work back up to your fork
 5. Submit a **Pull request** so that we can review your changes

NOTE: Be sure to merge the latest from "upstream" before making a pull request!


# License
Apache 2.0

CHANGELOG.md
# 1.5
## Features
- Project rename


# 1.4
## Features
- Add pods list data API endpoint
- Add pod logs data API endpoint


application/utils.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

"""
    @project: k8scontroller-metadata-exporter
    @component: core
    @author: vfabi
    @contact person: vfabi
    @support: vfabi
    @initial date: 2020-05-17 18:30:40
    @license: this file is subject to the terms and conditions defined
        in file 'LICENSE', which is part of this source code package
    @description
        Small utils.
    @todo

    THIS FILE IS PART OF @PROJECT_NAME@ PROJECT.
    PLEASE DO NOT MODIFY THIS FILE.
"""

def strictNamespaceMappingEnvarParse(envar=None):
    """Custom format STRICT_NAMESPACE_MAPPING env variable parsing.

    Note:
        STRICT_NAMESPACE_MAPPING value example: 'frontend.develop.example.com:develop,frontend.staging.example.com:staging'
        No spaces or special chars.
    """

    if envar:
        items = envar.split(',')
        mapdict = {}
        for item in items:
            mapdict[item.split(':')[0]] = item.split(':')[1]
        return mapdict
    return None


app.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

"""
    @project: k8scontroller-metadata-exporter
    @component: core
    @author: vfabi
    @support: vfabi
    @initial date: 2020-05-17 18:30:40
    @license: this file is subject to the terms and conditions defined
        in file 'LICENSE', which is part of this source code package
    @description
        k8scontroller-metadata-exporter application entrypoint.
    @todo

    THIS FILE IS PART OF @PROJECT_NAME@ PROJECT.
    PLEASE DO NOT MODIFY THIS FILE.
"""

import os
from urllib.parse import urlparse
from kubernetes import client, config, watch
from flask import Flask, jsonify, request
from application.utils import strictNamespaceMappingEnvarParse


strict_namespece_mapping = strictNamespaceMappingEnvarParse(os.getenv('STRICT_NAMESPACE_MAPPING', None))
app = Flask(__name__)
config.load_incluster_config()


@app.route('/deployments/', methods=['GET'])
def deployments():
    """Get deployments meta and status data.

    Note:
        For filtering return data by namespace you have 2 options:
        1 - you can use mapping from STRICT_NAMESPACE_MAPPING env variable to strictly map request domain name to namespace. In this case
            second option will be passed.
        2 - you can use URL filter with 'namespace' argument, for example: http://www.alpha.example.com/deployments/?namespace=develop
            If request domain is specified in strict namespace mapping (STRICT_NAMESPACE_MAPPING env variable) this filtering feature won't work.

    Returns:
        json object:
            [
                {
                    "name": "testapp",
                    "namespace": "develop",
                    "images":[
                        "mcr.microsoft.com/aks/hcp/tunnel-openvpn:1.0.8",
                        "mcr.microsoft.com/aks/hcp/hcp-tunnel-front:v1.9.2-v3.0.14"
                    ],
                    "labels": {
                        "app": "testapp"
                    },
                    "status": {
                        "available_replicas": 1,
                        "ready_replicas": 1,
                        "replicas": 1,
                        "unavailable_replicas": null,
                        "updated_replicas": 1
                    }
                },
            ]
    """

    api = client.AppsV1Api()
    data = []
    domain = urlparse(request.url_root).netloc.split(':')[0]

    if strict_namespece_mapping:
        namespace = strict_namespece_mapping.get(domain, None)
        if not namespace:
            namespace = request.args.get('namespace')
    else:
        namespace = request.args.get('namespace')

    try:
        deployments = api.list_deployment_for_all_namespaces(watch=False)
        for deployment in deployments.items:
            deploymentdata = {}
            deploymentdata['images'] = []
            if namespace == deployment.metadata.namespace:
                deploymentdata['name'] = deployment.metadata.name
                deploymentdata['namespace'] = deployment.metadata.namespace
                deploymentdata['labels'] = deployment.metadata.labels
                for container in deployment.spec.template.spec.containers:
                    deploymentdata['images'].append(container.image)
                deploymentdata['status'] = {
                    'replicas': deployment.status.replicas,
                    'available_replicas': deployment.status.available_replicas,
                    'ready_replicas': deployment.status.ready_replicas,
                    'unavailable_replicas': deployment.status.unavailable_replicas,
                    'updated_replicas': deployment.status.updated_replicas
                }
                data.append(deploymentdata)
            if not namespace:
                deploymentdata['name'] = deployment.metadata.name
                deploymentdata['namespace'] = deployment.metadata.namespace
                deploymentdata['labels'] = deployment.metadata.labels
                for container in deployment.spec.template.spec.containers:
                    deploymentdata['images'].append(container.image)
                deploymentdata['status'] = {
                    'replicas': deployment.status.replicas,
                    'available_replicas': deployment.status.available_replicas,
                    'ready_replicas': deployment.status.ready_replicas,
                    'unavailable_replicas': deployment.status.unavailable_replicas,
                    'updated_replicas': deployment.status.updated_replicas
                }
                data.append(deploymentdata)
        return jsonify(data)
    except Exception as e:
        response = {'error': f'Exception. Details: {e}.'}
        return jsonify(response), 500


@app.route('/pods/', methods=['GET'])
def pods():
    """Get pods meta and status data.

    Returns:
        json object:
            [
                {
                    "name": "testapp-423hag423a41-6wha",
                    "namespace": "develop",
                    "pod_ip": "1.1.1.23",
                    "node_ip": "100.100.1.84",
                    "node_name": "awshost-host-100.100.1.84",
                    "status": "Running",
                    "start_time": datetime.datetime(2020, 6, 23, 20, 30, 55, tzinfo=tzlocal()),
                    "labels": {
                        "app": "testapp",
                    },
                },
            ]
    """

    api = client.CoreV1Api()
    data = []
    domain = urlparse(request.url_root).netloc.split(':')[0]

    if strict_namespece_mapping:
        namespace = strict_namespece_mapping.get(domain, None)
        if not namespace:
            namespace = request.args.get('namespace')
    else:
        namespace = request.args.get('namespace')

    try:
        pods = api.list_pod_for_all_namespaces(watch=False)
        for pod in pods.items:
            poddata = {}
            if namespace == pod.metadata.namespace:
                poddata['name'] = pod.metadata.name
                poddata['pod_ip'] = pod.status.pod_ip
                poddata['node_ip'] = pod.status.host_ip
                poddata['node_name'] = pod.spec.node_name
                poddata['namespace'] = pod.metadata.namespace
                poddata['status'] = pod.status.phase
                poddata['start_time'] = pod.status.start_time
                poddata['labels'] = pod.metadata.labels
                data.append(poddata)
            if not namespace:
                poddata['name'] = pod.metadata.name
                poddata['pod_ip'] = pod.status.pod_ip
                poddata['node_ip'] = pod.status.host_ip
                poddata['node_name'] = pod.spec.node_name
                poddata['namespace'] = pod.metadata.namespace
                poddata['status'] = pod.status.phase
                poddata['start_time'] = pod.status.start_time
                poddata['labels'] = pod.metadata.labels
                data.append(poddata)
        return jsonify(data)

    except Exception as e:
        response = {'error': f'Exception. Details: {e}.'}
        return jsonify(response), 500


@app.route('/pod/logs/', methods=['GET'])
def pod_logs():
    """Get pod logs data."""

    api = client.CoreV1Api()
    domain = urlparse(request.url_root).netloc.split(':')[0]

    if strict_namespece_mapping:
        namespace = strict_namespece_mapping.get(domain, None)
        if not namespace:
            namespace = request.args.get('namespace')
        else:
            if namespace != request.args.get('namespace'):
                response = {'error': 'Access to this namespace denied. STRICT_NAMESPACE_MAPPING enabled.'}
                return jsonify(response), 400
    else:
        namespace = request.args.get('namespace')
    pod = request.args.get('pod')
    tail_lines = request.args.get('tail_lines')

    if not pod or not namespace or not tail_lines:
        response = {'error': 'Incorrect request arguments.'}
        return jsonify(response), 400

    try:
        logs = api.read_namespaced_pod_log(
            name=pod,
            namespace=namespace,
            tail_lines=int(tail_lines),
            timestamps=True
        )
        return jsonify(logs)
    except Exception as e:
        response = {'error': f'Exception. Details: {e}.'}
        return jsonify(response), 500


if __name__ == '__main__':
    app.run(
        debug=False,
        host='0.0.0.0',
        port=8000
    )
